res/controllers/Traktor-Kontrol-S3-hid-scripts.js

```diff
@@ -0,0 +1,2227 @@
+};
+
+// colorOutput drives lights that have the palettized multicolor lights.
+TraktorS3.Channel.prototype.colorOutput = function(value, key) {
+    var ledValue = this.channelBaseColor();
+    if (value === 1 || value === true) {
+        ledValue += TraktorS3.LEDBrightValue;
+    } else {
+        ledValue += TraktorS3.LEDDimValue;
+    }
+    this.controller.hid.setOutput(this.group, key, ledValue, !this.controller.batchingOutputs);
+};
+
+TraktorS3.Channel.prototype.hotcuesOutput = function(_value, group, key) {
+    var deck = this.controller.Channels[group].parentDeck;
+    if (deck.activeChannel !== group) {
+        // Not active, ignore
+        return;
+    }
+    var matches = key.match(/hotcue_(\d+)_/);
+    if (matches.length !== 2) {
+        HIDDebug("Didn't get expected hotcue number from string: " + key);
+        return;
+    }
+    var cueNum = matches[1];
+    deck.lightHotcue(cueNum);
+};
+
+// Returns true if there was an update.
+TraktorS3.Channel.prototype.lightWheelPosition = function() {
+    if (!this.positionUpdated) {
+        return false;
+    }
+    this.positionUpdated = false;
+    var rotations = this.curPosition * (1 / 1.8);  // 1/1.8 is rotations per second
+    // Calculate angle from 0-1.0
+    var angle = rotations - Math.floor(rotations);
+    // The wheel has 8 segments
+    var wheelAngle = 8.0 * angle;
+    var baseLedValue = this.channelBaseColor();
+    // Reduce the dimming distance at the end of track.
+    var dimDistance = this.endOfTrack ? 2.5 : 1.5;
+    var segValues = [0, 0, 0, 0, 0, 0, 0, 0];
+    for (var seg = 0; seg < 8; seg++) {
+        var distance = TraktorS3.wheelSegmentDistance(seg, wheelAngle);
+        var brightVal = Math.round(4 * (1.0 - (distance / dimDistance)));
+        if (this.endOfTrack) {
+            dimDistance = 1.5;
+            brightVal = Math.round(4 * (1.0 - (distance / dimDistance)));
+            if (this.endOfTrackBlinkState) {
+                brightVal = brightVal > 0x03 ? 0x04 : 0x02;
+            } else {
+                brightVal = brightVal > 0x02 ? 0x04 : 0x00;
+            }
+        }
+        if (brightVal <= 0) {
+            segValues[seg] = 0x00;
+        } else {
+            segValues[seg] = baseLedValue + brightVal - 1;
+        }
+    }
+    this.parentDeck.wheelOutput(this.group, segValues);
+    return true;
+};
+
+// FXControl is an object that manages the gray area in the middle of the
+// controller: the fx control knobs, fxenable buttons, and fx select buttons.
+TraktorS3.FXControl = function(controller) {
+    // 0 is filter, 1-4 are FX Units 1-4
+    this.FILTER_EFFECT = 0;
+    this.activeFX = this.FILTER_EFFECT;
+    this.controller = controller;
+
+    this.enablePressed = {
+        "[Channel1]": false,
+        "[Channel2]": false,
+        "[Channel3]": false,
+        "[Channel4]": false
+    };
+    this.selectPressed = {
+        0: false,
+        1: false,
+        2: false,
+        3: false,
+        4: false
+    };
```

Comment:
Why not use an Array instead of an Object for this?

Reply:
probably leftover from some old code.  fixed